File: iPhone/VerBank-CSTSv3-IOSClientApi/ClientStation/api/doc/operator/user/UserDocOperator.m
//
//  UserDocOperator.m
//  VerBank-CSTSv3-IOSClientApi
//
//  Created By Zhanglei on 15/7/1.
//  Copyright (c) 2015年 allone. All rights reserved.
//

#import "UserDocOperator.h"
#import "IP_TRADESERV5018.h"
#import "OP_TRADESERV5018.h"

#import "IP_TRADESERV5017.h"
#import "OP_TRADESERV5017.h"

#import "IP_TRADESERV5019.h"
#import "OP_TRADESERV5019.h"

#import "IP_TRADESERV5022.h"
#import "OP_TRADESERV5022.h"

#import "IP_TRADESERV5024.h"
#import "OP_TRADESERV5024.h"

#import "IP_TRADESERV5025.h"
#import "OP_TRADESERV5025.h"

#import "CSTSCaptain.h"
#import "APIDoc.h"

static UserDocOperator *userDocOperator = nil;

@implementation UserDocOperator

+ (UserDocOperator *)getInstance{
    if (userDocOperator == nil) {
        userDocOperator = [[UserDocOperator alloc] init];
    }
    return userDocOperator;
}

- (Boolean)loadUserLogin:(NSString *)aeid{
//        ListenerCaptain.getApiDataEventListener().onLoadUserLogin(APIDataEventListener.STATE_START);
    
    IP_TRADESERV5018 *ip = [[IP_TRADESERV5018 alloc] init];
    [ip setUserName:aeid];
    OPFather *opFather = [[CSTSCaptain getInstance] trade:ip];
    if (![opFather isSuccessed]) {
//        ListenerCaptain.getApiDataEventListener().onLoadUserLogin(APIDataEventListener.STATE_FAILED);
        return false;
    }
    
    OP_TRADESERV5018 *op = (OP_TRADESERV5018 *)opFather;
    [[APIDoc getUserDocCaptain] addUserLogin:[op getUserLogin]];
    
//    ListenerCaptain.getApiDataEventListener().onLoadUserLogin(APIDataEventListener.STATE_SUCCEED);
    return true;
}

- (Boolean)loadGroupConfig:(NSString *)groupName{
//        ListenerCaptain.getApiDataEventListener().onLoadGroupConfig(APIDataEventListener.STATE_START);
    
    IP_TRADESERV5017 *ip = [[IP_TRADESERV5017 alloc] init];
    [ip setGroupName:groupName];
    OPFather *opFather = [[CSTSCaptain getInstance] trade:ip];
    if (![opFather isSuccessed]) {
//        ListenerCaptain.getApiDataEventListener().onLoadGroupConfig(APIDataEventListener.STATE_FAILED);
        return false;
    }

    OP_TRADESERV5017 *op = (OP_TRADESERV5017 *)opFather;
    [[APIDoc getUserDocCaptain] addGroups:[op getGroupConfigArray]];
    
//        ListenerCaptain.getApiDataEventListener().onLoadGroupConfig(APIDataEventListener.STATE_SUCCEED);
    return true;
}

- (Boolean)loadAccountStrategy:(long long)accountID{
    return [self loadAccountStrategyWitAccountID:accountID aeid:nil];
}

- (Boolean)loadAccountStrategyWitAccountID:(long long)accountID aeid:(NSString *)aeid{
//        ListenerCaptain.getApiDataEventListener().onLoadAccountStrategy(APIDataEventListener.STATE_START);
    
    IP_TRADESERV5019 *ip = [[IP_TRADESERV5019 alloc] init];
    [ip setAccountID:accountID];
    OPFather *opFather = [[CSTSCaptain getInstance] trade:ip];
    if (![opFather isSuccessed]) {
//        ListenerCaptain.getApiDataEventListener().onLoadAccountStrategy(APIDataEventListener.STATE_FAILED);
        return false;
    }
    
    OP_TRADESERV5019 *op = (OP_TRADESERV5019 *)opFather;
    AccountStrategy *accountStrategy = [op getAccountStrategy];
    if (aeid != nil && (![[accountStrategy getAeid] isEqualToString:aeid])) {
        return false;
    }
    [[APIDoc getUserDocCaptain] setAccountStrategy:[op getAccountStrategy]];
    
//        ListenerCaptain.getApiDataEventListener().onLoadAccountStrategy(APIDataEventListener.STATE_SUCCEED);
    return true;
}

- (Boolean)loadMoneyAccount:(long long)accountID{
//        ListenerCaptain.getApiDataEventListener().onLoadMoneyAccount(APIDataEventListener.STATE_START);
    IP_TRADESERV5022 *ip = [[IP_TRADESERV5022 alloc] init];
    [ip setAccountID:accountID];
    OPFather *opFather = [[CSTSCaptain getInstance] trade:ip];
    if (![opFather isSuccessed]) {
//        ListenerCaptain.getApiDataEventListener().onLoadMoneyAccount(APIDataEventListener.STATE_FAILED);
        return false;
    }
    OP_TRADESERV5022 *op = (OP_TRADESERV5022 *)opFather;
    [[APIDoc getUserDocCaptain] setMoneyAccount:[op getMoneyAccount]];
    if ([APIDoc isInited]) {
        JEDI_SYS_Try {
            [[APIDoc getFixUtil] fixAccount:[[APIDoc getUserDocCaptain] getCDS_AccountStore:accountID]
                            ForceToScanTrde:true];
        }JEDI_SYS_EndTry
    }

//        ListenerCaptain.getApiDataEventListener().onLoadMoneyAccount(APIDataEventListener.STATE_SUCCEED);
//        ListenerCaptain.getApiDataChangeListener().onMoneyAccountChanged(new Long(accountID));
    NSString *account  = [NSString stringWithFormat:@"%lld", accountID];
    [API_IDEventCaptain fireEventChanged:DATA_ON_MoneyAccount_Changed eventData:account];
        return true;
}

- (Boolean)loadTrade4CFD:(long long)accountID{
//        ListenerCaptain.getApiDataEventListener().onLoadTrade4CFD(APIDataEventListener.STATE_START);
    IP_TRADESERV5024 *ip = [[IP_TRADESERV5024 alloc] init];
    [ip setAccountID:accountID];
    OPFather *opFather = [[CSTSCaptain getInstance]trade:ip];
    if (![opFather isSuccessed]) {
//        ListenerCaptain.getApiDataEventListener().onLoadTrade4CFD(APIDataEventListener.STATE_FAILED);
        return false;
    }
    OP_TRADESERV5024 *op = (OP_TRADESERV5024 *)opFather;
    [[APIDoc getUserDocCaptain] resetTradeByAccount:accountID Trades:[op getTTradeArray]];
    
    if ([APIDoc isInited]) {
        JEDI_SYS_Try{
            [[APIDoc getFixUtil] fixAccount:[[APIDoc getUserDocCaptain] getCDS_AccountStore:accountID]
                            ForceToScanTrde:true];
        }JEDI_SYS_EndTry
    }
//        ListenerCaptain.getApiDataEventListener().onLoadTrade4CFD(APIDataEventListener.STATE_SUCCEED);
//        API_IDEventCaptain.getInstance().fireEventChanged(new API_IDEvent(API_IDEvent_NameInterface.DATA_ON_Trade_Changed, new Long[] { accountID }, true, ""));
//        ListenerCaptain.getApiDataChangeListener().onTradeChanged(new Long(accountID));
    NSString *account = [NSString stringWithFormat:@"%lld", accountID];
    [API_IDEventCaptain fireEventChanged:DATA_ON_Trade_Changed eventData:account];
        return true;
}

- (Boolean)loadOrder4CFD:(long long)accountID{
//        ListenerCaptain.getApiDataEventListener().onLoadOrder4CFD(APIDataEventListener.STATE_START);
    IP_TRADESERV5025 *ip = [[IP_TRADESERV5025 alloc] init];
    [ip setAccountID:accountID];
    OPFather *opFather = [[CSTSCaptain getInstance] trade:ip];
    if (![opFather isSuccessed]) {
        return false;
    }
    OP_TRADESERV5025 *op = (OP_TRADESERV5025 *)opFather;
    [[APIDoc getUserDocCaptain] resetOrderByAccount:accountID TOrderArray:[op getTOrderArray]];
    [API_IDEventCaptain fireEventChanged:DATA_ON_Order_Changed eventData:[@(accountID) stringValue]];
    
    return true;
}

@end

File: WeChatService/EX02-CSTSv3-ClientAPI/src/jedi/ex01/client/station/api/doc/operator/user/UserDocOperator.java
package jedi.ex01.client.station.api.doc.operator.user;

import jedi.ex01.CSTS3.comm.ipop.IP_TRADESERV5017;
import jedi.ex01.CSTS3.comm.ipop.IP_TRADESERV5018;
import jedi.ex01.CSTS3.comm.ipop.IP_TRADESERV5019;
import jedi.ex01.CSTS3.comm.ipop.IP_TRADESERV5022;
import jedi.ex01.CSTS3.comm.ipop.IP_TRADESERV5024;
import jedi.ex01.CSTS3.comm.ipop.IP_TRADESERV5025;
import jedi.ex01.CSTS3.comm.ipop.OPFather;
import jedi.ex01.CSTS3.comm.ipop.OP_TRADESERV5017;
import jedi.ex01.CSTS3.comm.ipop.OP_TRADESERV5018;
import jedi.ex01.CSTS3.comm.ipop.OP_TRADESERV5019;
import jedi.ex01.CSTS3.comm.ipop.OP_TRADESERV5022;
import jedi.ex01.CSTS3.comm.ipop.OP_TRADESERV5024;
import jedi.ex01.CSTS3.comm.ipop.OP_TRADESERV5025;
import jedi.ex01.CSTS3.comm.struct.AccountStrategy;
import jedi.ex01.client.station.api.CSTS.CSTSCaptain;
import jedi.ex01.client.station.api.doc.DataDoc;
import jedi.ex01.client.station.api.doc.DocCaptain;
import jedi.ex01.client.station.api.doc.util.AccountScanUtil;
import jedi.ex01.client.station.api.event.API_IDEventCaptain;
import jedi.ex01.client.station.api.event.API_IDEvent_NameInterface;

public class UserDocOperator {
	private static UserDocOperator userDocOperator = new UserDocOperator();

	private UserDocOperator() {
	}

	public boolean loadUserLogin(String aeid) {
		IP_TRADESERV5018 ip = new IP_TRADESERV5018();
		ip.setUserName(aeid);
		OPFather opfather = CSTSCaptain.getInstance().trade(ip);
		if (!opfather.isSucceed()) {
			return false;
		}
		OP_TRADESERV5018 op = (OP_TRADESERV5018) opfather;
		DataDoc.getInstance().setUserLogin(op.getUserLogin());
		return true;
	}

	public boolean loadGroupConfig(String groupName) {
		IP_TRADESERV5017 ip = new IP_TRADESERV5017();
		ip.setGroupName(groupName);
		OPFather opfather = CSTSCaptain.getInstance().trade(ip);
		if (!opfather.isSucceed()) {
			return false;
		}
		OP_TRADESERV5017 op = (OP_TRADESERV5017) opfather;
		DataDoc.getInstance().setGroup(op.getGroupConfigVec()[0]);
		return true;
	}

	public boolean loadAccountStrategy(long accountID) {
		return loadAccountStrategy(accountID, null);
	}

	public boolean loadAccountStrategy(long accountID, String aeid) {
		IP_TRADESERV5019 ip = new IP_TRADESERV5019();
		ip.setAccountID(accountID);
		OPFather opfather = CSTSCaptain.getInstance().trade(ip);
		if (!opfather.isSucceed()) {
			return false;
		}
		OP_TRADESERV5019 op = (OP_TRADESERV5019) opfather;
		AccountStrategy accountStrategy = op.getAccountStrategy();
		if (aeid != null && (!accountStrategy.getAeid().equals(aeid))) {
			return false;
		}
		DataDoc.getInstance().getAccountStore().setStrategy(op.getAccountStrategy());
		return true;
	}

	public boolean loadMoneyAccount(long accountID) {
		IP_TRADESERV5022 ip = new IP_TRADESERV5022();
		ip.setAccountID(accountID);
		OPFather opfather = CSTSCaptain.getInstance().trade(ip);
		if (!opfather.isSucceed()) {
			return false;
		}
		OP_TRADESERV5022 op = (OP_TRADESERV5022) opfather;
		DataDoc.getInstance().getAccountStore().resetMoneyAccounts(op.getMoneyAccountVec());
		if (DocCaptain.isInited()) {
			try {
				AccountScanUtil.fixAccount(true);
			} catch (Exception e) {
				e.printStackTrace();
			}
		}
		API_IDEventCaptain.fireEventChanged(API_IDEvent_NameInterface.DATA_ON_MoneyAccount_Changed);
		return true;
	}

	public boolean loadTrade4CFD(long accountID) {
		IP_TRADESERV5024 ip = new IP_TRADESERV5024();
		ip.setAccountID(accountID);
		OPFather opfather = CSTSCaptain.getInstance().trade(ip);
		if (!opfather.isSucceed()) {
			return false;
		}
		OP_TRADESERV5024 op = (OP_TRADESERV5024) opfather;
		DataDoc.getInstance().resetTrades(op.getTrade4CFDVec());

		if (DocCaptain.isInited()) {
			try {
				AccountScanUtil.fixAccount(true);
			} catch (Exception e) {
				e.printStackTrace();
			}
		}
		API_IDEventCaptain.fireEventChanged(API_IDEvent_NameInterface.DATA_ON_Trade_Changed);
		return true;
	}

	public boolean loadOrder4CFD(long accountID) {
		IP_TRADESERV5025 ip = new IP_TRADESERV5025();
		ip.setAccountID(accountID);
		OPFather opfather = CSTSCaptain.getInstance().trade(ip);
		if (!opfather.isSucceed()) {
			return false;
		}
		OP_TRADESERV5025 op = (OP_TRADESERV5025) opfather;
		DataDoc.getInstance().resetOrders(op.getOrders4CFDVec());
		API_IDEventCaptain.fireEventChanged(API_IDEvent_NameInterface.DATA_ON_Order_Changed);
		return true;
	}

	public static UserDocOperator getUserDocOperator() {
		return userDocOperator;
	}
}
